### ISSUE TITLE
self type on property creates a spurious error

### ISSUE THREAD

**Issue by JelleZijlstra** [python/mypy#3223]
Running mypy master on this file:
```python
from typing import TypeVar, List
_T = TypeVar('_T')

class X:
    @property
    def __members__(self: _T) -> List[_T]: ...

reveal_type(X().__members__)
```
gives
```
../bin/selftypeproperty.py:8: error: Revealed type is 'builtins.list[selftypeproperty.X*]'
../bin/selftypeproperty.py:8: error: Invalid method type
```
The revealed type is correct but the "Invalid method type" doesn't make sense.

**Comment by gvanrossum** [comment]
Is it the same if you rename `__members__` to `foo`?

**Comment by JelleZijlstra** [comment]
Yes, doesn't make a difference.

**Comment by elazarg** [comment]
I should look into it.

Note that adding a bound, i.e. `_T = TypeVar('_T', bound='X')` seems to fix this.

**Comment by JelleZijlstra** [comment]
Thanks! FYI, the code where I found this had a `Type` over a bound typevar (python/typeshed#1195). It has `_T = TypeVar('_T', bound=Enum)` and the property is `def __members__(self: Type[_T]) -> Mapping[str, _T]: ...`.
